### PR TITLE
[RFC] Add method to DDP to check for backward finalization.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Summary: In cases where DDP backward is not finalized, the error is raised only in the next forward iteration of DDP. However, if there are other collective calls between those two points, training scripts could potentially get stuck.

As a result, there should be a way to check if DDP finalized after calling `.backward()`. To address this, I've added a `_check_reducer_finalized` method to validate that DDP indeed did successfully finish reduction.

Test Plan: Added unit tests.

### CODE REVIEW

File: torch/testing/_internal/distributed/distributed_test.py [hunk 1]
```diff
@@ -9722,6 +9722,39 @@ def forward(self, inp):
...
+            input = torch.rand(10, 10, requires_grad=True).cuda(self.rank)
+            ddp = torch.nn.parallel.DistributedDataParallel(
+                model,
+                device_ids=[self.rank],
+            )
+
+            out1, _ = ddp(input)
+            out1.sum().backward()
+
+            with self.assertRaisesRegex(RuntimeError, "Reducer has not been finalized!"):
```
Comment (rohan-varma) [inline torch/testing/_internal/distributed/distributed_test.py:9751]: if this is run with find_unused_parameters=True or static_graph (and the graph is static) then I believe this training should work. In that case, will _check_finalized return true or false?

File: torch/nn/parallel/distributed.py [hunk 2]
```diff
@@ -2235,4 +2235,22 @@ def _set_static_graph(self):
...
+        """
+        Checks if the reducer has processed all buckets and finalized the backward
+        appropriately.
+
+        It is useful to call this method after calling .backward() in your trianing loop
+        in order to avoid subsequent hard to debug errors down the road due to the
+        reducer not finalizing backward.
+        """
+        if not self.reducer._has_finalized():
+            raise RuntimeError(
```
Comment (rohan-varma) [inline torch/nn/parallel/distributed.py:2253]: should it return True or False or raise an exception as we're doing here?

File: torch/csrc/distributed/c10d/reducer.cpp [hunk 3]
```diff
@@ -2239,4 +2239,8 @@ void Reducer::remove_autograd_hooks() {
   hooks_.clear();
 }
 
+bool Reducer::has_finalized() {
+  return !require_finalize_;
```
Comment (rohan-varma) [inline torch/csrc/distributed/c10d/reducer.cpp:2243]: I think we have a lock around this access, do we need that here?

**Review by rohan-varma — COMMENTED** [review]
thanks for the PR!

Trying to understand how stuckness can happen. Is the issue that one rank is somehow not finalized and other ranks have finalized, and move on to other collectives, resulting in a hang? Is it possible to have an example of this?

**Comment by pritamdamania87** [comment]
> Trying to understand how stuckness can happen. Is the issue that one rank is somehow not finalized and other ranks have finalized, and move on to other collectives, resulting in a hang? Is it possible to have an example of this?

Yes this is how the stuckness happens. You can think of different ranks training different "task heads" of a model and resulting in some ranks finalizing and others not since the set of parameters used on different ranks are different. The unit test in the PR can easily be updated such that only certain ranks use the additional loss and others don't. Happy to update the unit test accordingly to illustrate this

File: torch/testing/_internal/distributed/distributed_test.py (hunk 1 above)
Comment (pritamdamania87) [reply torch/testing/_internal/distributed/distributed_test.py:9751]: Good point, the issue I ran into was with `find_unused_parameters=True`. Let me update the unit test with `find_unused_parameters` to illustrate the problem.

File: torch/csrc/distributed/c10d/reducer.cpp (hunk 3 above)
Comment (pritamdamania87) [reply torch/csrc/distributed/c10d/reducer.cpp:2243]: Good catch!

File: torch/testing/_internal/distributed/distributed_test.py (hunk 1 above)
Comment (pritamdamania87) [reply torch/testing/_internal/distributed/distributed_test.py]: @rohan-varma Updated the test

File: torch/csrc/distributed/c10d/reducer.cpp (hunk 3 above)
Comment (rohan-varma) [reply torch/csrc/distributed/c10d/reducer.cpp:2243]: Need to add the lock?

File: torch/testing/_internal/distributed/distributed_test.py [hunk 4]
```diff
@@ -9708,6 +9712,58 @@ def forward(self, inp):
...
+            input = torch.rand(10, 1024, requires_grad=True).cuda(self.rank)
+            ddp = torch.nn.parallel.DistributedDataParallel(
+                model,
+                device_ids=[self.rank],
+                find_unused_parameters=True,
+                bucket_cap_mb=(1024 * 4 / 1024 / 1024),  # One bucket per parameter.
+            )
+
+            if self.rank == 0:
+                out1, _ = ddp(input)
```
Comment (rohan-varma) [inline torch/testing/_internal/distributed/distributed_test.py:9749]: oh ok. I thought that DDP would be able to support this, as this is "not all outputs used in loss". But it seems that the custom class causes DDP to have an issue.

**Review by rohan-varma — APPROVED** [review]
LGTM, thanks!

**Comment by rohan-varma** [comment]
@pritamdamania87 I'll merge after CI is cleared

File: torch/csrc/distributed/c10d/reducer.cpp (hunk 3 above)
Comment (pritamdamania87) [reply torch/csrc/distributed/c10d/reducer.cpp:2243]: Ah I thought I added it in, will fix it

**Comment by pritamdamania87** [comment]
@rohan-varma Looks like CI is green now.

**Review by rohan-varma — APPROVED** [review]

**Review by rohan-varma — APPROVED** [review]
